app/root.tsx

```diff
@@ -40,34 +45,81 @@ export const meta: MetaFunction = () => ({
 export interface RootLoaderData {
   ENV: ReturnType<typeof getClientEnv>
   user: Awaited<Auth0Profile | undefined>
+  endpoints: EndpointsQuery | null
 }
 
 export const loader: LoaderFunction = async ({ request }) => {
   const user = await authenticator.isAuthenticated(request)
 
-  const data = {
-    ENV: getClientEnv(),
-    user: user?.profile,
+  let endpointsResponse
+  let portalUserId
+
+  if (user) {
+    const portal = initPortalClient({ token: user?.accessToken })
+    endpointsResponse = await portal.endpoints().catch((e) => {
+      console.log(e)
+    })
+
+    const getPortalUserIdResponse = await portal.getPortalUserID().catch((e) => {
+      console.log(e)
+    })
+
+    portalUserId = getPortalUserIdResponse?.getPortalUserID
+
+    if (!portalUserId && user) {
+      const email = user?.profile?._json?.email
+      const providerUserID = user?.profile?.id
+
+      invariant(email, "email is not found")
+      invariant(providerUserID, "providerUserID is not found")
+
+      const portalAdmin = await initAdminPortal(portal)
+
+      await portalAdmin.adminCreatePortalUser({
+        email,
+        providerUserID,
+      })
+    }
   }
 
-  return json<RootLoaderData>(data)
+  return json<RootLoaderData>({
+    ENV: getClientEnv(),
+    user: user?.profile,
+    endpoints: endpointsResponse ?? null,
+  })
 }
 
 createEmotionCache({ key: "pni" })
 
+export type GlobalOutletContext = {
+  endpoints: EndpointsQuery | null
+  portalUserId: string | undefined
+}
+
 export default function App() {
-  const { ENV, user } = useLoaderData<RootLoaderData>()
+  const { ENV, user, endpoints } = useLoaderData<RootLoaderData>()
   const { isPlasmic } = useRoot({ user })
+  const shouldRenderApp = useMemo(
+    () => Boolean(endpoints) && Boolean(user),
+    [endpoints, user],
+  )
   return (
     <>
       {isPlasmic ? (
         <PlasmicContainer />
       ) : (
         <RootProviders>
           <Document>
-            <main style={{ padding: 0 }}>
-              <Outlet />
-            </main>
+            {shouldRenderApp ? (
+              <RootAppShell
+                endpoints={endpoints as EndpointsQuery}
+                user={user as Auth0Profile}
+              >
+                <Outlet />
+              </RootAppShell>
+            ) : (
+              <div> Loading ...</div>
```

Comment:
Loading is an entire topic I never found a specific pattern i liked for it. This is fine for now, but we should discuss a pattern for that.

Reply:
Yup, this is just a placeholder for now